// lib_wehaus/index.js
'use strict';

const rest = require('restler'), EventEmitter = require('events');

class Wehaus extends EventEmitter {

  constructor(config) {
    super();
    const defaults = { 
      host: 'http://dev.patagoniclabs.com'
    }
    this.config = Object.assign(defaults, config);
  }

  get_token(email, password) {
    let this_ = this;

    rest.postJson( 
        this.config.host + '/users/sign_in',
        {user: {email: email, password: password}},
        {headers: this.default_headers()} )

      .on('success',function(data, response){
        this_.config.token = data.authentication_token;
        this_.config.email = email;
        console.log('Got token '+this_.config.token+' for email '+this_.config.email);
        this_.emit('ready'); })

      .on('fail',this.catch_error);
  }

  get_mac_address() {
    const exec = require('child_process').exec;
    let this_ = this;
    let do_with_mac = function(e, out, err) {
      if (e) {
        console.error(`exec error: ${e}`);
        return;
      }
      out = out.trim();
      console.log("Got Mac Address "+out);
      this_.config.mac = out;
    }

    exec("ifconfig | grep -Po \"(HWaddr|ether) [a-f0-9:]+\" | head -n 1 | cut -d ' ' -f 2 | sed \"s/://g\"", do_with_mac);
  }

  auth_headers() {
    if ( typeof this.config.token === 'undefined' ) {
      throw new Error('No token configured');
    } else {
      return { "X-User-Email": this.config.email, "X-User-Token": this.config.token };
    }
  }

  default_headers() {
    return {'Accept': 'application/json'};
  }

  full_headers() {
    return Object.assign(this.auth_headers(), this.default_headers());
  }

  send_value(device, value) {
    rest.putJson(
        this.api_url('device/' + device.get_api_id() + '/data'),
        {value: value},
        {headers: this.full_headers()})

      .on('success', function(data, response) {
        console.log('success'); })

      .on('fail',this.catch_error);
  }

  get_devices() {
    let this_ = this;
    rest.get(this.api_url('devices'), 
        {headers: this.full_headers()})

      .on('success', (data, response) => {
        this_.devices = data.filter((d) => {
          return (d.remote_status.address == this_.config.mac);
        }).map((d) => {
          return { id: d.id, primitive: d.primitive, endpoint: d.endpoint, status: d.text_status }; 
        });
        console.log('Got device list',this.devices);
        this_.emit('ready'); })

      .on('fail',this.catch_error);
  }

  catch_error(e) {
    console.error(e);
  }

  api_url(path) {
    return this.config.host + '/api/v2/' + path;
  }

  get_device_by_endpoint(endpoint) {
    let d = this.devices.filter((d) => {
      return (d.endpoint == endpoint);
    });
    if (d.length == 0) {
      console.error('Endpoint does not exist');
    } else {
      return d.shift();
    }
    return false;
  }
}

class Device extends EventEmitter {
  constructor(w, endpoint) {
    super();
    this.Wehaus = w;
    this.endpoint = endpoint;
    let d = this.Wehaus.get_device_by_endpoint(endpoint);
    if ( !d ) { throw new Error('Device not found'); }
    console.log('Found device: ', d);
    this.id = d.id;
    this.device = d;
  }

  get_api_id() {
    return this.id;
  }
}

class DataDevice extends Device {
  constructor(w,e) {
    super(w,e);
    this.on('data',this.send_value);
    this.primitive = 'DataDevice';
    if ( this.primitive != this.device.primitive ) { throw new Error('Device has the wrong type for the endpoint'); }
  }

  send_value(value) {
    this.Wehaus.send_value(this, value);
  }
}

class SensorDevice extends DataDevice {}

class AlarmDevice extends DataDevice {}


class ActionDevice extends Device {
  constructor(w,e) {
    super(w,e);
    this.primitive = 'ActionDevice';
    if ( this.primitive != this.device.primitive ) { throw new Error('Device has the wrong type for the endpoint'); }
  }
}

class ToggableDevice extends ActionDevice {
  on() {
  }

  off() {
  }
}

class LevelDevice extends ToggableDevice {
  constructor(w, e) {
    const min_level = 0, max_level = 255;
    super(w, e);
  }

  on() {
    set_level(max_level);
  }

  off() {
    set_level(min_level);
  }

  set_level(level) {
  }
}

module.exports.Wehaus = Wehaus;
module.exports.SensorDevice = SensorDevice;
